### ISSUE TITLE
Allow for use with abstract classes and mixins

### ISSUE THREAD

**Issue by stergiom** [Bonemind/syringe#6]
The following two examples seem like they should work, is this something that worked with earlier versions for Crystal?

```cr
abstract class Foo
  abstract def foo(*args)
end

class A < Foo
  def call(arg)
    puts "A#call with #{arg}"
  end
end

class B
  def initialize(@a : Foo)
    @a.call("<arg from B>")
  end
end

B.new(A.new) # A#call with <arg from B>

Syringe.injectable(A)
Syringe.wrap(B) # Error: undefined method 'getFoo' for Syringe:Module

B.new
```

```cr
module Foo
  abstract def foo(*args)
end

class A
  include Foo

  def call(arg)
    puts "A#call with #{arg}"
  end
end

class B
  def initialize(@a : Foo)
    @a.call("<arg from B>")
  end
end

B.new(A.new) # A#call with <arg from B>

Syringe.injectable(A)
Syringe.wrap(B) # Error: undefined method 'getFoo' for Syringe:Module

B.new

**Comment by elliotize** [comment]
@stergiom So I think that this logic would be strange due to what would happen if multiple classes included the mixin or abstract class. I added a feature the allows you to use an array to handle this situation.
Try this
```
module Foo
  abstract def foo(*args)
end

class A
  include Foo

  def call(arg)
    puts "A#call with #{arg}"
  end
end

class B
  def initialize(@a : Array(Foo))
    @a.first.call("<arg from B>")
  end
end

B.new(A.new) # A#call with <arg from B>

Syringe.injectable(A)
Syringe.wrap(B) # Error: undefined method 'getFoo' for Syringe:Module

B.new
```

**Comment by stergiom** [comment]
Thanks @elliotize this does work, but for using mixins as one does interfaces, you're left having to wrap/unwrap a single element array, which for this use-case seems unnecessary.

I don't think strange logic is required for cases where multiple _concrete implementations_ (to stick with interface terminology) are made available, Syringe could raise an error or warning (picking the first/last/preferred implementation) to deal with the ambiguity.

From the readme; https://github.com/Bonemind/syringe#use-case is not too far off, perhaps for easier unit testing or the flexibility of providing multiple IoC container scenarios, something like this might be useful;

```cr
# i_db_client
module App
  module IDbClient
    # ..
  end
end

# spec_helper.cr
class TestDbClient
  include App::IDbClient
  # ..
end

# app_db_client.cr
module App
  class AppDbClient
    include App::IDbClient
    # ..
  end
end

# other_db_client.cr
module App
  class OtherDbClient
    include App::IDbClient
    # ..
  end
end

# foo_controller.cr
module App
  class FooController
    def initialize(@db_client : IDbClient)
    end
  end
end

# prod_iocc.cr
Syringe.injectable(App::AppDbClient)
Syringe.wrap(App::FooController)

# staging_iocc.cr
Syringe.injectable(App::OtherDbClient)
Syringe.wrap(App::FooController)
```

**Comment by elliotize** [comment]
@stergiom yes this can be done easily then

**Comment by elliotize** [comment]
@stergiom before I started down this road of DI I was using https://github.com/waterlink/mocks.cr. I wonder if we can enabled an easy way to use it with Syringe instead of adding duplicate features. At the same time I still think what your are trying to do here is possible and I will put up a PR ASAP.

**Comment by stergiom** [comment]
Fair point to using mocks with DI, together they do make testing easier, but I feel that this addresses one specific facet of using DI; how to best write tests. Setting aside the topic of whether there's value to adding `mocks.cr` specific functionality, I think there's more to look at on the DI container front that could justify the duplication of functionality.

DI on its own does make mocking easier, but leaves you having to make adjustments to the architecture of your app/lib for deciding what will instantiate/deliver the dependencies. `Syringe` offers an implicit way of doing this via mixing the relative functionality in, but this doesn't do much for loosening coupling up (which is a different facet that DI is generally used for).

To gain this win, `Syringe` allows for `#injectable` and `#wrap` to be invoked independently of the aspects you're looking to inject/wrap/(or both). Now you've got more control/flexibility, but still have to deal with what will invoke the coupling. DI/IoC containers emerge as a logical next step here and (arguably) their use is justified by the value that decoupling offers an evolving project.

The last issue to deal with is that even though aspects are being coupled independently (implying flexibility), with `Syringe` they are still bound by Type. Being able to depend on more abstract constructs allows for easier evolution as illustrated by the [Dependency Inversion Principle](https://en.wikipedia.org/wiki/Dependency_inversion_principle), in the case of Crystal, being able to use a mixin as a _kind_ of interface brings DI, DI/IoC-containers and DIP together in a way (of-course not the only way) that allows for more graceful growth of a project, and it's in the least an approach that has worked for me.

`Syringe` in context I see as a light DI/DI-container toolkit/lib that should allow for this flexibility, without the last bit —allowing for concrete classes to depend on abstract classes/modules, it comes in one step short.

My apologies for the wall-of-text, if interested in a more detailed study on the subject, [DIP in the wild](https://martinfowler.com/articles/dipInTheWild.html) is the best write-up I know.

**Comment by elliotize** [comment]
I think I know what your asking for I tried to create a system to auto wrap ancestors given a criteria or order and it worked. However I felt it was hacky and the developer could easily shoot themselves in the foot. I decided on a new approach where the user can specify injection of any class. I created this https://github.com/Bonemind/syringe/pull/7/commits/f992a703aaf3088452fcb94f6dcfbfb8369b9a0a hopefully it addresses this feature your asking for.

**Comment by elliotize** [comment]
I also created a PR however it still contains my other PR that has not merged yet https://github.com/Bonemind/syringe/pull/7

**Comment by stergiom** [comment]
Nice! `injectable_as` seems sensible, thanks @elliotize, I'll test this out later in the day. From a brief look, this might break the proc functionality added with https://github.com/elliotize/syringe/commit/7e567a22310edd681316ae9cab44d61d68e6a17a#diff-834d556b2f6b96ef2bd4fd21518ccaed if so, do you have a branch that accommodates a `Proc(Thing)` too, this way I can test it out against an ongoing project that makes use of modules, procs and now injectable-as.

**Comment by stergiom** [comment]
Just fiddled a variant with proc functionality in, works a treat, thanks. I'll test it against something more convoluted soon. @Bonemind, I think this PR covers the earlier work, adds flexibility with `injectable_as` and comments through the steps, how do you feel about merging?
